### DISCUSSION TITLE
Using `retryEither` with suspend functions?

### DISCUSSION THREAD

**Discussion by milgner** [arrow-kt/arrow#3406] [Q&A] [answered]
Current version: arrow-resilience 1.2.4

Trying to convert some code that loads data over the network to retry in case of failure using `arrow-resilience`.
The existing code is a `suspend` function that returns an `Either` with the result or an error.

Now I noticed that `retryEither` is declared as

```
public suspend inline fun <Error, Result, Output> Schedule<Error, Output>.retryEither(
  @BuilderInference action: () -> Either<Error, Result>,
): Either<Error, Result>
```

But since my existing function is a `suspend` function, I can't use it for the `action` argument.

So I'm looking for a way to combine this. Ideally it would work the same with a signature like

```
public suspend inline fun <Error, Result, Output> Schedule<Error, Output>.retryEither(
  @BuilderInference action: suspend () -> Either<Error, Result>,
): Either<Error, Result>
```

**Comment by nomisRev** [comment]
Hey @milgner,

I think this is some kind-of mistake in the API. `action` _should_ support through `inline` and `suspend` keyword on the main function, but apparently not the case so we need to explicitly specify `suspend`.

I think this can be considered a bug, and if you're interested I'd be great to have a ticket, or a PR for this!

**Reply by milgner** [reply]
Oh, alright! Thank you for the quick response! Since some of the other callbacks were already marked as `suspend`, I assumed that it was intentionally left out for `retryEither`. I'll gladly open a PR to change it :+1:

**Reply by nomisRev** [reply]
Awesome, a PR would be wonderful! Thank you ☺️

**Reply by milgner** [reply]
Forgot about this as for some reason it started working again.

**Comment by kyay10** [comment, marked as answer]
I think the reason this started working again is maybe you were trying to use a reference instead of a lambda? `retryEither(::someSuspendFunction)` would not work, annoyingly. However, `retryEither { someSuspendFunction() }` absolutely works as long as you're already in a `suspend` context.
